File: src/integration/spring/src/jvmMain/java/community/flock/wirespec/integration/spring/java/web/WirespecResponseBodyAdvice.java
package community.flock.wirespec.integration.spring.java.web;

import community.flock.wirespec.integration.spring.shared.RawJsonBody;
import community.flock.wirespec.java.Wirespec;
import org.springframework.core.MethodParameter;
import org.springframework.http.HttpStatusCode;
import org.springframework.http.MediaType;
import org.springframework.http.converter.HttpMessageConverter;
import org.springframework.http.server.ServerHttpRequest;
import org.springframework.http.server.ServerHttpResponse;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.servlet.mvc.method.annotation.ResponseBodyAdvice;

import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

@ControllerAdvice
public class WirespecResponseBodyAdvice implements ResponseBodyAdvice<Object> {

    private final Wirespec.Serialization wirespecSerialization;
    private final Map<Class<?>, Method> toResponseCache = new ConcurrentHashMap<>();

    public WirespecResponseBodyAdvice(Wirespec.Serialization wirespecSerialization) {
        this.wirespecSerialization = wirespecSerialization;
    }

    @Override
    public boolean supports(MethodParameter returnType, Class<? extends HttpMessageConverter<?>> converterType) {
        return Wirespec.Response.class.isAssignableFrom(returnType.getParameterType());
    }

    @Override
    @SuppressWarnings("unchecked")
    public Object beforeBodyWrite(
            Object body,
            MethodParameter returnType,
            MediaType selectedContentType,
            Class<? extends HttpMessageConverter<?>> selectedConverterType,
            ServerHttpRequest request,
            ServerHttpResponse response
    ) {
        try {
            Class<?> declaringClass = returnType.getParameterType().getDeclaringClass();
            Method toResponse = toResponseCache.computeIfAbsent(declaringClass, cls -> {
                Class<?> handlerClass = Arrays.stream(cls.getDeclaredClasses())
                        .filter(c -> c.getSimpleName().equals("Handler"))
                        .findFirst()
                        .orElseThrow(() -> new IllegalStateException("Handler not found in " + cls));
                return Arrays.stream(handlerClass.getDeclaredMethods())
                        .filter(m -> m.getName().equals("toResponse") && Modifier.isStatic(m.getModifiers()))
                        .findFirst()
                        .orElseThrow(() -> new IllegalStateException("toResponse method not found in " + handlerClass));
            });

            if (body instanceof Wirespec.Response<?> wirespecResponse) {
                Wirespec.RawResponse rawResponse = (Wirespec.RawResponse) toResponse.invoke(null, wirespecSerialization, wirespecResponse);
                
                response.setStatusCode(HttpStatusCode.valueOf(rawResponse.statusCode()));
                for (Map.Entry<String, List<String>> entry : rawResponse.headers().entrySet()) {
                    response.getHeaders().put(entry.getKey(), entry.getValue());
                }
                return rawResponse.body().map(RawJsonBody::new).orElse(null);
            }
            return body;
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
}

File: src/integration/spring/src/jvmMain/kotlin/community/flock/wirespec/integration/spring/kotlin/web/WirespecResponseBodyAdvice.kt
package community.flock.wirespec.integration.spring.kotlin.web

import community.flock.wirespec.integration.spring.shared.RawJsonBody
import community.flock.wirespec.kotlin.Wirespec
import org.springframework.core.MethodParameter
import org.springframework.core.io.Resource
import org.springframework.http.HttpStatusCode
import org.springframework.http.MediaType
import org.springframework.http.converter.HttpMessageConverter
import org.springframework.http.server.ServerHttpRequest
import org.springframework.http.server.ServerHttpResponse
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.servlet.mvc.method.annotation.ResponseBodyAdvice
import java.lang.reflect.Method
import java.util.concurrent.ConcurrentHashMap

@ControllerAdvice
class WirespecResponseBodyAdvice(
    private val wirespecSerialization: Wirespec.Serialization,
) : ResponseBodyAdvice<Any?> {

    private val toResponseCache = ConcurrentHashMap<Class<*>, Method>()

    override fun supports(returnType: MethodParameter, converterType: Class<out HttpMessageConverter<*>?>): Boolean = Wirespec.Response::class.java.isAssignableFrom(returnType.parameterType)

    @Suppress("UNCHECKED_CAST")
    override fun beforeBodyWrite(
        body: Any?,
        returnType: MethodParameter,
        selectedContentType: MediaType,
        selectedConverterType: Class<out HttpMessageConverter<*>>,
        request: ServerHttpRequest,
        response: ServerHttpResponse,
    ): Any? {
        val declaringClass = returnType.parameterType.declaringClass
        val toResponse = toResponseCache.computeIfAbsent(declaringClass) { cls ->
            cls.declaredMethods.first { it.name == "toResponse" }
        }
        val instance = declaringClass.getDeclaredField("INSTANCE").get(null)
        return when (body) {
            is Wirespec.Response<*> -> {
                val rawResponse = toResponse.invoke(instance, wirespecSerialization, body) as Wirespec.RawResponse
                response.setStatusCode(HttpStatusCode.valueOf(rawResponse.statusCode))
                response.headers.putAll(rawResponse.headers)
                val responseBody = body.body
                if (responseBody is Resource) {
                    if (!response.headers.containsKey("Content-Type")) {
                        response.headers.set("Content-Type", MediaType.APPLICATION_OCTET_STREAM_VALUE)
                    }
                    responseBody.inputStream.use { it.copyTo(response.body) }
                    response.body.flush()
                    null
                } else if (rawResponse.body == null) {
                    Unit
                } else {
                    rawResponse.body?.let { RawJsonBody(it) }
                }
            }
            else -> body
        }
    }
}
